# Locating image files on S3 and understanding the path structure

The full JUMP-CP image data is hosted in a public bucket: `s3://cellpainting-gallery/cpg0016-jump/`. The goal of this notebook is to find where the images of my three selected plates live in that bucket, and to understand how the hierarchy (source, batch, plate, well, imaging site, fluorescence channel) is reflected in the file paths.

No image is downloaded here; this notebook only manipulates metadata and S3 file listings.

In [ ]:
# Install s3fs so pandas can read S3 directly (anonymous, public bucket)
!pip install s3fs --quiet

import pandas as pd
import s3fs

# Reuse the working subset selected in notebook 02
SUBSET_SOURCE = "source_3"
SUBSET_BATCH  = "CP59"
SUBSET_PLATES = ["BR5867a3", "BR5867b3", "BR5867c3"]

# Anonymous filesystem handle for the public S3 bucket
fs = s3fs.S3FileSystem(anon=True)
BUCKET = "cellpainting-gallery"

print(f"Subset: {SUBSET_SOURCE}, batch {SUBSET_BATCH}, plates {SUBSET_PLATES}")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 203.4/203.4 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.5/89.5 kB 5.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.0/15.0 MB 66.4 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datasets 4.0.0 requires fsspec[http]<=2025.3.0,>=2023.1.0, but you have fsspec 2026.4.0 which is incompatible.
gcsfs 2025.3.0 requires fsspec==2025.3.0, but you have fsspec 2026.4.0 which is incompatible.
Subset: source_3, batch CP59, plates ['BR5867a3', 'BR5867b3', 'BR5867c3']


In [ ]:
# What's at the root of the JUMP-CP dataset on S3?
fs.ls(f"{BUCKET}/cpg0016-jump/")[:20]

['cellpainting-gallery/cpg0016-jump/source_1',
 'cellpainting-gallery/cpg0016-jump/source_10',
 'cellpainting-gallery/cpg0016-jump/source_11',
 'cellpainting-gallery/cpg0016-jump/source_13',
 'cellpainting-gallery/cpg0016-jump/source_15',
 'cellpainting-gallery/cpg0016-jump/source_2',
 'cellpainting-gallery/cpg0016-jump/source_3',
 'cellpainting-gallery/cpg0016-jump/source_4',
 'cellpainting-gallery/cpg0016-jump/source_5',
 'cellpainting-gallery/cpg0016-jump/source_6',
 'cellpainting-gallery/cpg0016-jump/source_7',
 'cellpainting-gallery/cpg0016-jump/source_8',
 'cellpainting-gallery/cpg0016-jump/source_9',
 'cellpainting-gallery/cpg0016-jump/source_all']

In [ ]:
fs.ls(f"{BUCKET}/cpg0016-jump/{SUBSET_SOURCE}/")

['cellpainting-gallery/cpg0016-jump/source_3/images',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace_dl']

In [ ]:
fs.ls(f"{BUCKET}/cpg0016-jump/{SUBSET_SOURCE}/workspace/")

['cellpainting-gallery/cpg0016-jump/source_3/workspace/analysis',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/backend',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv_orig',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/profiles']

In [ ]:
fs.ls(f"{BUCKET}/cpg0016-jump/{SUBSET_SOURCE}/workspace/load_data_csv/")[:20]

['cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP60',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_25_all_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_26_all_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_27_all_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_28_all_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_29_all_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_31_all_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_32_all_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_33_all_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP_34_mix_Phenix1',
 'cellpainting-gallery/cpg0016-jump/source_3/wo

In [ ]:
fs.ls(f"{BUCKET}/cpg0016-jump/{SUBSET_SOURCE}/workspace/load_data_csv/{SUBSET_BATCH}/")[:20]

['cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867a3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867b3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867c3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867d3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5868a3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5868b3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5868c3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5868d3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5869a3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5869b3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5869c3',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_

In [ ]:
plate_dir = f"{BUCKET}/cpg0016-jump/{SUBSET_SOURCE}/workspace/load_data_csv/{SUBSET_BATCH}/{SUBSET_PLATES[0]}/"
fs.ls(plate_dir)

['cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867a3/load_data.csv',
 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867a3/load_data_with_illum.csv']

In [ ]:
# Take whichever load_data file is in the directory
load_data_files = [f for f in fs.ls(plate_dir) if "load_data" in f and f.endswith(".csv")]
print(f"Files found: {load_data_files}")

# Pick the simplest one (without illumination correction info)
load_data_path = "s3://" + [f for f in load_data_files if "illum" not in f][0]
print(f"Loading: {load_data_path}")

load_data = pd.read_csv(load_data_path, storage_options={"anon": True})
print(f"Shape: {load_data.shape}")
print(f"Columns: {list(load_data.columns)}")
load_data.head()

Files found: ['cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867a3/load_data.csv', 'cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867a3/load_data_with_illum.csv']
Loading: s3://cellpainting-gallery/cpg0016-jump/source_3/workspace/load_data_csv/CP59/BR5867a3/load_data.csv
Shape: (3456, 34)
Columns: ['URL_OrigRNA', 'URL_OrigMito', 'URL_OrigER', 'URL_OrigDNA', 'URL_OrigAGP', 'Metadata_Source', 'Metadata_Batch', 'Metadata_Plate', 'Metadata_Well', 'Metadata_Site', 'Group_Number', 'Group_Index', 'Series_OrigMito', 'Series_OrigAGP', 'Series_OrigRNA', 'Series_OrigER', 'Series_OrigDNA', 'Frame_OrigMito', 'Frame_OrigAGP', 'Frame_OrigRNA', 'Frame_OrigER', 'Frame_OrigDNA', 'Channel_OrigMito', 'Channel_OrigAGP', 'Channel_OrigRNA', 'Channel_OrigER', 'Channel_OrigDNA', 'Metadata_ChannelNumber', 'Metadata_FileLocation', 'Metadata_Frame', 'Metadata_Series', 'Metadata_Row', 'Metadata_Col', 'Metadata_ArrayGroup']


,URL_OrigRNA,URL_OrigMito,URL_OrigER,URL_OrigDNA,URL_OrigAGP,Metadata_Source,Metadata_Batch,Metadata_Plate,Metadata_Well,Metadata_Site,...,Channel_OrigRNA,Channel_OrigER,Channel_OrigDNA,Metadata_ChannelNumber,Metadata_FileLocation,Metadata_Frame,Metadata_Series,Metadata_Row,Metadata_Col,Metadata_ArrayGroup
0,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,source_3,CP59,BR5867a3,A01,1,...,-1,-1,-1,NaN,NaN,0,0,1,1,1
1,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,source_3,CP59,BR5867a3,A01,2,...,-1,-1,-1,NaN,NaN,0,0,1,1,1
2,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,source_3,CP59,BR5867a3,A01,3,...,-1,-1,-1,NaN,NaN,0,0,1,1,1
3,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,source_3,CP59,BR5867a3,A01,4,...,-1,-1,-1,NaN,NaN,0,0,1,1,1
4,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,s3://cellpainting-gallery/cpg0016-jump/source_...,source_3,CP59,BR5867a3,A01,5,...,-1,-1,-1,NaN,NaN,0,0,1,1,1


In [ ]:
# How many imaging sites per well?
sites_per_well = load_data.groupby("Metadata_Well")["Metadata_Site"].nunique()
print(f"Number of wells in this plate: {len(sites_per_well)}")
print(f"Sites per well (typical values):")
print(sites_per_well.value_counts())

Number of wells in this plate: 384
Sites per well (typical values):
Metadata_Site
9    384
Name: count, dtype: int64


In [ ]:
# Inspect what a single row contains: one well, one site, multiple channels
single_row = load_data.iloc[0]
print("Identifying columns:")
print(single_row[["Metadata_Plate", "Metadata_Well", "Metadata_Site"]])
print()
print("Image file paths and names:")
for col in load_data.columns:
    if col.startswith("PathName_Orig") or col.startswith("FileName_Orig"):
        print(f"  {col}: {single_row[col]}")

Identifying columns:
Metadata_Plate    BR5867a3
Metadata_Well          A01
Metadata_Site            1
Name: 0, dtype: object

Image file paths and names:


In [ ]:
# What image-related columns does this load_data file actually have?
image_cols = [c for c in load_data.columns if any(p in c for p in ["Path", "File", "URL", "Channel"])]
print("Image-related columns found in this load_data:")
for c in image_cols:
    print(f"  {c}")

Image-related columns found in this load_data:
  URL_OrigRNA
  URL_OrigMito
  URL_OrigER
  URL_OrigDNA
  URL_OrigAGP
  Channel_OrigMito
  Channel_OrigAGP
  Channel_OrigRNA
  Channel_OrigER
  Channel_OrigDNA
  Metadata_ChannelNumber
  Metadata_FileLocation


In [ ]:
# For one specific well/site, show the S3 path to each of the 5 fluorescence channels

url_cols = sorted([c for c in load_data.columns if c.startswith("URL_Orig")])

print(f"Identifying info:")
print(f"  Plate: {single_row['Metadata_Plate']}")
print(f"  Well:  {single_row['Metadata_Well']}")
print(f"  Site:  {single_row['Metadata_Site']}")
print()
print("Full S3 paths to each fluorescence channel at this position:")
for col in url_cols:
    channel = col.replace("URL_Orig", "")
    print(f"  {channel:5s} -> {single_row[col]}")

Identifying info:
  Plate: BR5867a3
  Well:  A01
  Site:  1

Full S3 paths to each fluorescence channel at this position:
  AGP   -> s3://cellpainting-gallery/cpg0016-jump/source_3/images/CP59/images/BR5867a3__2022-04-29T01_02_16-Measurement 1/Images/r01c01f01p01-ch2sk1fk1fl1.tiff
  DNA   -> s3://cellpainting-gallery/cpg0016-jump/source_3/images/CP59/images/BR5867a3__2022-04-29T01_02_16-Measurement 1/Images/r01c01f01p01-ch5sk1fk1fl1.tiff
  ER    -> s3://cellpainting-gallery/cpg0016-jump/source_3/images/CP59/images/BR5867a3__2022-04-29T01_02_16-Measurement 1/Images/r01c01f01p01-ch4sk1fk1fl1.tiff
  Mito  -> s3://cellpainting-gallery/cpg0016-jump/source_3/images/CP59/images/BR5867a3__2022-04-29T01_02_16-Measurement 1/Images/r01c01f01p01-ch1sk1fk1fl1.tiff
  RNA   -> s3://cellpainting-gallery/cpg0016-jump/source_3/images/CP59/images/BR5867a3__2022-04-29T01_02_16-Measurement 1/Images/r01c01f01p01-ch3sk1fk1fl1.tiff


## Summary

### The structure of JUMP-CP on S3

The JUMP-CP data is in the public AWS bucket `s3://cellpainting-gallery/cpg0016-jump/`. At the top, I find one folder per data source: `source_1` through `source_15`, with `source_12` and `source_14` missing as expected from the metadata, plus a `source_all` folder that I have not investigated yet.

Inside each source, three folders:

- `images/`: the actual TIFF files
- `workspace/`: CellProfiler outputs, metadata tables, and the "load_data" index files that I need
- `workspace_dl/`: deep learning related outputs, most likely pre-computed feature embeddings (worth coming back to later)

### Where image paths are listed

The index files live in `workspace/load_data_csv/<batch>/<plate>/`. For my batch `CP59` I find 20+ plates including my 3 selected ones (`BR5867a3`, `BR5867b3`, `BR5867c3`). Each plate folder contains two files: `load_data.csv` (the basic index) and `load_data_with_illum.csv` (same index plus illumination correction file paths, which I do not need yet).

### What's inside a load_data.csv

For plate `BR5867a3`, the `load_data.csv` has **3,456 rows and 34 columns**. The 3,456 rows correspond exactly to **384 wells × 9 imaging sites per well**. Every well of this plate is imaged at 9 different positions by the microscope, so I will need to handle this 9-fold replication every time I work on a single well.

Each row represents one (well, site) combination, with these key columns:

- `Metadata_Plate`, `Metadata_Well`, `Metadata_Site`: identifying the row
- 5 columns `URL_OrigXXX` (one per Cell Painting fluorescence channel): each one gives the full S3 path to one image file

The 5 channels of the Cell Painting assay are: **DNA** (nuclei), **RNA**, **ER** (endoplasmic reticulum), **AGP** (actin / golgi / plasma membrane), **Mito** (mitochondria). One (well, site) combination produces 5 image files, one per channel.

A note on column naming convention: in `source_3`, each image is referenced by a single `URL_OrigXXX` column that contains the full S3 path. Other sources of JUMP-CP can use a `PathName_*` + `FileName_*` pair instead. The column convention is therefore worth checking when changing source.

### How the image files are named

A full S3 path looks like this:

s3://cellpainting-gallery/cpg0016-jump/source_3/images/CP59/images/BR5867a3__2022-04-29T01_02_16-Measurement 1/Images/r01c01f01p01-ch5sk1fk1fl1.tiff

Breaking this down:

- The plate folder name includes the acquisition timestamp: `BR5867a3__2022-04-29T01_02_16-Measurement 1`. So plate BR5867a3 was imaged on 29 April 2022.
- The filename `r01c01f01p01-ch5sk1fk1fl1.tiff` encodes the position compactly:
  - `r01c01` = row 1, column 1 of the plate, i.e. well A01
  - `f01` = imaging field (site) number 1
  - `p01` = z-plane 1 (Cell Painting is 2D imaging, so always 1)
  - `ch5` = channel number 5

An important subtlety: the channel number in the filename does NOT correspond directly to the channel name listed in the CSV column. The mapping I observed for this plate is:

| Filename code | Channel name |
|:-------------:|:------------:|
| ch1           | Mito         |
| ch2           | AGP          |
| ch3           | RNA          |
| ch4           | ER           |
| ch5           | DNA          |

So it is safer to always use the `URL_OrigXXX` column from the load_data CSV than to try to parse the filename. The CSV column already tells me which channel is which.
